packages/runtime-common/realm.ts

```diff
@@ -524,13 +559,214 @@ export class Realm {
+    let statuses = errors
+      .filter((e) => e.status)
+      .map((e) => e.status) as number[];
+    return Math.min(...statuses);
+  }
+
+  private async handleAtomicOperations(
+    request: Request,
+    requestContext: RequestContext,
+  ): Promise<Response> {
+    let body = await request.text();
+    let json;
+    try {
+      json = JSON.parse(body);
+    } catch (e) {
+      return createResponse({
+        body: JSON.stringify({
+          errors: [
+            {
+              title: 'Invalid atomic:operations format',
+              detail: `Request body is not valid JSON`,
+            },
+          ],
+        }),
+        init: {
+          status: 400,
+          headers: {
+            'content-type': SupportedMimeType.JSONAPI,
+          },
+        },
+        requestContext,
+      });
+    }
+    let validationErrors = this.validate(json);
+    if (validationErrors.length > 0) {
+      return createResponse({
+        body: JSON.stringify({ errors: validationErrors }),
+        init: {
+          status: 400,
+          headers: { 'content-type': SupportedMimeType.JSONAPI },
+        }, //consolidate to 400
+        requestContext,
+      });
+    }
+    let operations = filterAtomicOperations(json['atomic:operations']);
+    let atomicCheckErrors = await this.checkBeforeAtomicWrite(operations);
+    if (atomicCheckErrors.length > 0) {
+      return createResponse({
+        body: JSON.stringify({ errors: atomicCheckErrors }),
+        init: {
+          status: this.lowestStatusCode(atomicCheckErrors),
+          headers: { 'content-type': SupportedMimeType.JSONAPI },
+        },
+        requestContext,
+      });
+    }
+    let files = new Map<LocalPath, string>();
+    let writeResults: FileWriteResult[] = [];
+
+    for (let operation of operations) {
+      let resource = operation.data;
+      let href = operation.href;
+
+      let fileURL = this.paths.fileURL(href);
+      let localPath = this.paths.local(fileURL);
+      if (isModuleResource(resource)) {
+        files.set(localPath, resource.attributes?.content ?? '');
+      } else if (isCardResource(resource)) {
+        let doc = {
+          data: resource,
+        };
+        files.set(localPath, JSON.stringify(doc, null, 2));
+      } else {
+        return createResponse({
+          body: JSON.stringify({
+            errors: [
+              {
+                status: 400,
+                title: 'Invalid resource',
+                detail: `Operation data is not a valid card resource or module resource`,
+              },
+            ],
+          }),
+          init: {
+            status: 400,
+            headers: { 'content-type': SupportedMimeType.JSONAPI },
+          },
+          requestContext,
+        });
+      }
+    }
+
+    if (files.size > 0) {
+      try {
+        writeResults = await this.writeMany(files, {
+          clientRequestId: request.headers.get('X-Boxel-Client-Request-Id'),
+          serializeFile: true,
```

Comment:
its easier to handle `serializeFile` internally bcos if we read from the header, its really not that clear weather an `_atomic` write is providing source or card. So we just include an option `serializeFile` to instruct all card resources to be serialized.